File: ios/flutter_timezone/Sources/flutter_timezone/FlutterTimezonePlugin.m
#import "./include/flutter_timezone/FlutterTimezonePlugin.h"

@implementation FlutterTimezonePlugin

- (NSLocale *)getPreferredLocale {
  NSArray *preferredLanguages = [NSLocale preferredLanguages];
  if ([preferredLanguages count] > 0) {
    NSString *preferredIdentifier = [preferredLanguages firstObject];
    return [[NSLocale alloc] initWithLocaleIdentifier:preferredIdentifier];
  }
  return [NSLocale currentLocale];
}
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"flutter_timezone"
            binaryMessenger:[registrar messenger]];
  FlutterTimezonePlugin* instance = [[FlutterTimezonePlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (NSLocale *)preferredLocale {
  NSString *preferredIdentifier = NSLocale.preferredLanguages.firstObject;
  if (preferredIdentifier.length == 0) {
    return NSLocale.currentLocale;
  }

  return [[NSLocale alloc] initWithLocaleIdentifier:preferredIdentifier];
}

- (NSLocale *)localeFromArgument:(id)argument {
  if (![argument isKindOfClass:[NSString class]]) {
    return [self preferredLocale];
  }

  NSString *localeIdentifier = (NSString *)argument;
  if (localeIdentifier.length == 0) {
    return nil;
  }

  NSLocale *locale = [[NSLocale alloc] initWithLocaleIdentifier:localeIdentifier];
  NSString *languageCode = [locale objectForKey:NSLocaleLanguageCode];
  if (languageCode.length == 0) {
    return nil;
  }

  return locale;
}

- (NSDictionary *)timezoneInfoForTimeZone:(NSTimeZone *)timeZone locale:(NSLocale *)locale {
  NSString *localizedName = nil;
  NSString *localeIdentifier = nil;

  if (locale != nil) {
    localizedName = [timeZone localizedName:NSTimeZoneNameStyleStandard locale:locale];
    localeIdentifier = locale.localeIdentifier;
  }

  return @{
    @"identifier": timeZone.name,
    @"localizedName": localizedName ?: [NSNull null],
    @"locale": localeIdentifier ?: [NSNull null],
  };
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getLocalTimezone" isEqualToString:call.method]) {
    NSString *localeIdentifier = call.arguments;
    NSLocale *locale = nil;
    
    if (localeIdentifier != nil && [localeIdentifier isKindOfClass:[NSString class]]) {
      NSLocale *requestedLocale = [[NSLocale alloc] initWithLocaleIdentifier:localeIdentifier];
      // Check if the locale is valid by verifying it has a valid language code
      if ([requestedLocale languageCode] != nil) {
        locale = requestedLocale;
      } else {
        locale = nil;
      }
    } else {
      locale = [self getPreferredLocale];
    }
    
    NSTimeZone *timeZone = [NSTimeZone localTimeZone];
    NSString *localizedName = locale != nil ? [timeZone localizedName:NSTimeZoneNameStyleStandard locale:locale] : nil;
    
    result(@{
      @"identifier": [timeZone name],
      @"localizedName": localizedName,
      @"locale": [locale localeIdentifier]
    });
  } else if([@"getAvailableTimezones" isEqualToString:call.method]) {
    NSString *localeIdentifier = call.arguments;
    NSLocale *locale = nil;
    
    if (localeIdentifier != nil && [localeIdentifier isKindOfClass:[NSString class]]) {
      NSLocale *requestedLocale = [[NSLocale alloc] initWithLocaleIdentifier:localeIdentifier];
      // Check if the locale is valid by verifying it has a valid language code
      if ([requestedLocale languageCode] != nil) {
        locale = requestedLocale;
      } else {
        locale = nil;
      }
    } else {
      locale = [self getPreferredLocale];
    }
    
    NSArray *timezoneNames = [NSTimeZone knownTimeZoneNames];
    NSMutableArray *timezones = [[NSMutableArray alloc] initWithCapacity:[timezoneNames count]];
    
    for (NSString *timezoneName in timezoneNames) {
      NSTimeZone *timezone = [NSTimeZone timeZoneWithName:timezoneName];
      NSString *localizedName = locale != nil ? [timezone localizedName:NSTimeZoneNameStyleStandard locale:locale] : nil;
      
      [timezones addObject:@{
        @"identifier": timezoneName,
        @"localizedName": localizedName,
        @"locale": [locale localeIdentifier]
      }];
    }
    
    result(timezones);
  }
  else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: ios/flutter_timezone/Sources/flutter_timezone/FlutterTimezonePlugin.swift
import Flutter
import Foundation

public class FlutterTimezonePlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_timezone", binaryMessenger: registrar.messenger())
    let instance = FlutterTimezonePlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getLocalTimezone":
      let locale = locale(from: call.arguments)
      result(timezoneInfo(for: TimeZone.current, locale: locale))
    case "getAvailableTimezones":
      let locale = locale(from: call.arguments)
      let timezones: [[String: Any]] = TimeZone.knownTimeZoneIdentifiers.compactMap { identifier in
        guard let timeZone = TimeZone(identifier: identifier) else {
          return nil
        }
        return timezoneInfo(for: timeZone, locale: locale)
      }
      result(timezones)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func locale(from argument: Any?) -> Locale? {
    guard let localeIdentifier = argument as? String else {
      return Locale(identifier: Locale.preferredLanguages.first ?? Locale.current.identifier)
    }

    guard !localeIdentifier.isEmpty else {
      return nil
    }

    let locale = NSLocale(localeIdentifier: localeIdentifier)
    return locale.object(forKey: .languageCode) == nil ? nil : locale as Locale
  }

  private func timezoneInfo(for timeZone: TimeZone, locale: Locale?) -> [String: Any] {
    [
      "identifier": timeZone.identifier,
      "localizedName": locale.flatMap {
        timeZone.localizedName(for: .standard, locale: $0)
      } ?? NSNull(),
      "locale": locale?.identifier ?? NSNull(),
    ]
  }
}
